gravitee-am-test/specs/gateway/mfa.jest.spec.ts

```diff
@@ -0,0 +1,689 @@
+    emailFactor = await createEmailFactor(smtpResource, domain, accessToken);
+
+    recoveryCodeFactor = await createRecoveryCodeFactor(domain, accessToken);
+
+    const domainStarted = await startDomain(domain.id, accessToken);
+    expect(domainStarted).toBeDefined();
+    expect(domainStarted.id).toEqual(domain.id);
+    domain = domainStarted;
+
+    smsTestApp = await createSMSTestApp(smsFactor, domain, accessToken);
+    bruteForceTestApp = await createBruteForceTestApp(smsFactor, domain, accessToken, mfaChallengeAttemptsResetTime);
+    emailTestApp = await createEmailTestApp(emailFactor, domain, accessToken);
+    recoveryCodeTestApp = await createRecoveryCodeTestApp(smsFactor, recoveryCodeFactor, domain, accessToken);
+
+    /*
+    * it is intentional to call setTimeout after creating apps. Cannot avoid this timeout call.
+    * At this point I haven't found a function which is similar to retry until specific http code is returned.
+    * jest.retryTimes(numRetries, options) didn't applicable in this case.
+    * */
+    await new Promise((r) => setTimeout(r, 10000));
+
+    const result = await getWellKnownOpenIdConfiguration(domain.hrid).expect(200);
+    openIdConfiguration = result.body;
+});
+
+describe("MFA", () => {
+    describe('MFA rate limit test', () => {
+        let user1;
+
+        it("throw mfa_request_limit_exceed after 2 request", async () => {
+            const clientId = smsTestApp.settings.oauth.clientId;
+
+            user1 = await buildCreateAndTestUser(domain.id, accessToken, 1);
+            expect(user1).toBeDefined();
+
+            const authResponse = await initiateLoginFlow(clientId, openIdConfiguration, domain);
+            const postLogin = await login(authResponse, user1, clientId);
+
+            // Post login authentication
+            const authorize = await performGet(postLogin.headers['location'], '', {
+                'Cookie': postLogin.headers['set-cookie']
+            }).expect(302);
+
+            expect(authorize.headers['location']).toBeDefined();
+            expect(authorize.headers['location']).toContain(`${process.env.AM_GATEWAY_URL}/${domain.hrid}/mfa/enroll`);
+
+            const enrollMFA = await enrollSMSFactor(authorize, smsFactor, domain);
+            const authorize2 = await performGet(enrollMFA.headers['location'], '', {
+                'Cookie': enrollMFA.headers['set-cookie']
+            }).expect(302);
+
+            expect(authorize2.headers['location']).toBeDefined();
+            expect(authorize2.headers['location']).toContain(`${process.env.AM_GATEWAY_URL}/${domain.hrid}/mfa/challenge`);
+
+            /**
+             * The number of the get requests is based on the gateway gravtitee.yml 'mfa_rate' configuration
+             * These assertions will fail or need to be updated if 'mfa_rate' configuration is changed
+             */
+            const expectedCode = [200, 200, 302];
+            for (let i = 0; i < expectedCode.length; i++) {
```

Comment:
You can use a `forEach` here which would arguably be cleaner/less verbose.

Reply:
replaced with `for of` instead of `forEach` for a cleaner code. The anonymous function has to be a `async` in case of `forEach` and it makes this slightly difficult in this case.